spec.html

```diff
@@ -3815,6 +3838,17 @@ <h1>InitializeReferencedBinding ( _V_, _W_ )</h1>
           1. Return _base_.InitializeBinding(_V_.[[ReferencedName]], _W_).
         </emu-alg>
       </emu-clause>
+
+      <emu-clause id="sec-makeprivatereference" aoid="MakePrivateReference">
+        <h1>MakePrivateReference ( _baseValue_, _privateIdentifier_ )</h1>
+        <emu-alg>
+          1. Let _env_ be the running execution context's PrivateEnvironment.
+          1. Let _privateNameBinding_ be ! ResolveBinding(_privateIdentifier_, _env_).
+          1. Let _privateName_ be ? GetValue(_privateNameBinding_).
+          1. Assert: _privateIdentifier_ is a Private Name.
```

Comment:
This definitely is supposed to be `_privateName_`.

What's happening here (as I understand it, anyway) is that `#name`s are tracked using  the new PrivateEnvironment component of execution contexts, and the PrivateEnvironment components are DeclarativeEnvironment which associate strings with Private Names. The idea is that looking up `#name`s is basically a kind of scope lookup, so this PR reuses the existing machinery there.

The direct answer to your question is, this `GetValue` returns a Private Name because the preceding `ResolveBinding` returns a Reference Record whose `[[Base]]` (if not null) is the PrivateEnvironment associated with some class, and those environment records have Private Names stored in their bindings.

Specifically, PrivateEnvironment declarative environments are populated exclusively in step 4 of the evaluation semantics for `ClassElementName : PrivateIdentifier`:

```
4. If the binding for _privateIdentifier_ in _privateEnvRec_ is an uninitialized binding, then
  a. Let _privateName_ be NewPrivateName(_privateIdentifier_).
  b. Perform ! privateEnvRec.InitializeBinding(_privateIdentifier_, _privateName_).
```

I believe this actually all works. That said, I personally dislike reusing the environment record machinery here - even just the basic DeclarativeEnvironments have a lot of stuff which is not relevant to PrivateEnvironments. And there's no place where we operate on an environment record without knowing whether it's for a PrivateEnvironment or not, so there's no need to conflate the two.

I'll make a PR switching it over to a new type, so we can discuss it.

Reply:
Obviated by https://github.com/tc39/ecma262/pull/1668/commits/fa2a42430b32a0f074906f68d6d078fd09d8d8cc.